esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,373 @@
+        if (sst==SSTnone ||sst==SSTall)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid ScopeType(%s) in filterNestedScopeTypes",scopeType);
+
+        resultArray.append(sst);
+    }
+    resultArray.sort(funcCompareUnsigned);
+}
+
+void CWUDetails::processStatsIntoResponse(IArrayOf<IEspWUScope> & respScopes,
+                                          IConstWUStatisticIterator & statsIter,
+                                          unsigned __int64 & maxTimestamp)
+{
+    unsigned maxDepth = scopeFilter.getMaxDepth();
+    bool limitToIdList = ids.ordinality()>0;
+    bool returnSpecifyReturnAttribs = returnAttribList.ordinality()>0;
+    bool fitlerByScopeTypes = scopeTypesFilterList.ordinality();
+    bool filterNestedScopeTypes = nestedScopeTypes.ordinality()>0;
+    bool filterByAttributeValue = arrayAttributeFilter.ordinality();
+
+    DBGLOG("Id list count: %d", ids.ordinality());
+    DBGLOG("ReturnAttribute  list count: %d", returnAttribList.ordinality());
+    DBGLOG("filterNestedScopeTypes count: %d", nestedScopeTypes.ordinality());
+    DBGLOG("AttributeFilter count: %d", arrayAttributeFilter.ordinality());
+
+    IArrayOf<IEspWUResponseAttribute> attribs;
+    SCMStringBuffer previousScope, previousScopeType, previousId;
+    unsigned statsCount = 0, statsAdded = 0;
+    StringArray nestedScopeList;
+    ForEach(statsIter)
+    {
+        ++statsCount;
+        IConstWUStatistic & stat = statsIter.query();
+
+        SCMStringBuffer scope;
+        stat.getScope(scope);
+        StringBuffer id;
+        getScopeIdFromScopeName(id, scope.str());
+        StatisticScopeType scopeType = stat.getScopeType();
+
+        // Check if scope name has changed. If so, need to return the list attributes for previous scope
+        if (strcmp(previousScope.str(),scope.str())!=0)
+        {
+            if (previousScope.length()>0 && attribs.ordinality()>0)
+            {
+                Owned<IEspWUScope> respScope = buildRespScope(previousScope.str(), previousScopeType.str(), previousId.str(), attribs);
+                respScopes.append(*respScope.getClear());
+                attribs.clear();
+            }
+            previousScope.set(scope.str());
+            previousScopeType.set(queryScopeTypeName(scopeType));
+            previousId.set(id.str());
+        }
+
+        unsigned scopeDepth = getScopeDepth(scope.str());
+        bool foundInNestedScopeList = false;// Used to avoid adding an existing scope entry to nestedScopeList
+                                            // - True means entry was found in nestedScopeList
+                                            // - False means list wasn't searched or searched and not found
+        if (maxDepth && scopeDepth > maxDepth)
+        {
+            if (!nestedDepth || scopeDepth > maxDepth + nestedDepth)
+               continue;
+            foundInNestedScopeList = nestedScopeList.bSearch(scope.str(),funcCompareString);
```

Comment:
I would expect this logic to be in the code that detects the scope has changed - e.g.

if there is an actively matched scope, and if the newscope is more deeply nested than the parent, then ..., 
...
otherwise indicate the scope should be skipped.